src/probe.bpf.c

```diff
@@ -28,257 +28,520 @@ struct {
-  for (int i = 0; i < DNS_MAX_DOMAIN_LEN && read_cursor <= read_end_ptr && write_cursor < write_end_ptr; i++) {
-    if (this_segment_len == 0) {
-      this_segment_len = *(u8*) (read_cursor++);
+  if (segment_len & 0b11000000) {
 
-      if (this_segment_len == 0) {
-        *(u8*) (write_cursor++) = '\0';
-        return write_cursor - (void*) name;
-      }
+    if (segment_len & 0b00111111) {
+      bpf_printk("[dnssnoop] Protocol Error: Non-zero segment len after pointer marker: %d", segment_len);
+      return BPF_LOOP_BREAK;
+    }
 
-      *(u8*) (write_cursor++) = '.';
-    } else {
-      this_segment_len--;
-      *(u8*) (write_cursor++) = *(u8*) (read_cursor++);
+    // Read out segment offset byte.
+    u8* segment_offset_ptr = bpf_dynptr_slice(c->record->buf, c->record->offset, NULL, 1);
+    if (segment_offset_ptr == NULL) {
+      return BPF_LOOP_BREAK;
     }
+    c->record->offset++;
+    u8 segment_offset = *segment_offset_ptr;
+
+    c->success = true;
+    return BPF_LOOP_BREAK;
   }
 
-  *(u8*) write_cursor = '\0';
-  return -1;
+  // Adjust offset to next segment.
+  c->record->offset += segment_len;
+
+  return BPF_LOOP_CONTINUE;
 }
 
-SEC("tracepoint/net/net_dev_xmit")
-int trace_egress(struct trace_event_raw_net_dev_xmit* ctx)
+static bool jump_dns_name(struct cursor* record)
 {
-  void* data = NULL;
-  u32 data_len = 0;
-  void* sk_buff_addr = BPF_CORE_READ(ctx, skbaddr);
+  struct segment_ctx ctx = { NULL, record, false };
+  bpf_loop(DNS_MAX_SEGMENTS, jump_name_segment, (void*) &ctx, 0);
 
-  struct iphdr ip_header = {};
-  struct udphdr udp_header = {};
-
-  int index = 0;
-  void* buf = bpf_map_lookup_elem(&scratch_buf, &index);
-  if (!buf) {
-    return EXIT_FAILURE;
+  if (ctx.success) {
+    return 0;
+  } else {
+    return -1;
   }
+}
 
-  struct sk_buff* skb = buf;
-  bpf_probe_read_kernel(skb, sizeof(struct sk_buff), sk_buff_addr);
-  DECODE_PACKETS_UDP_SKB((*skb), ip_header, udp_header, data, data_len, true);
+////////////////////////////////////////////////////////////////////////////////
+// Read DNS Name
+////////////////////////////////////////////////////////////////////////////////
 
-  if (ip_header.version != 4) {
-    return EXIT_FAILURE;
+static int read_name_segment(u32 i, void* ctx)
+{
+  struct segment_ctx* c = ctx;
+  char buf[DNS_MAX_SEGMENT_LEN];
+
+  // Read out segment length byte.
+  u8* segment_len_ptr = bpf_dynptr_slice(c->record->buf, c->record->offset, NULL, 1);
+  if (segment_len_ptr == NULL) {
+    return BPF_LOOP_BREAK;
   }
+  c->record->offset++;
+  u8 segment_len = *segment_len_ptr;
 
-  if (udp_header.dest != bpf_htons(53)) {
-    return EXIT_FAILURE;
+  // A segment of 0 length, marks the end of this name.
+  if (segment_len == 0) {
+    c->success = true;
+    return BPF_LOOP_BREAK;
   }
 
-  struct dnshdr dns = {};
-  READ_FROM_PACKET(struct dnshdr, dns, data, data_len);
+  if (segment_len & 0b11000000) {
+    if (segment_len & 0b00111111) {
```

Comment:
theres a nice trick here. i believe we have a `ONES(n)` macro.

Then `~ONES(n)` does the other mask.

It basically does: `(1 << n) - 1`

Reply:
Eh, even reading your comment & comparing it to my code I feel like that requires my brain to jump through an extra hoop. I'd rather keep it this way since I think it's clearer this way. Reasonable?

Reply:
Oh. Its a common trick.

Reply:
it get better when you need like `1<<n - 1` and n > 43 or some large no